ci/engine.go

```diff
@@ -0,0 +1,16 @@
+package main
+
+import "dagger.io/dagger"
+
+func (e EngineTargets) Lint(ctx dagger.Context) (string, error) {
```

Comment:
Out of scope for this PR, but since we're in the process of radically changing the DX, I'm going to put it out there:

`dagger.Connect()` exists for historical reasons. Those reasons are less and less relevant, maybe we could reconsider having a `Client` and doing a `Connect`?

Today:

```go
c, err := dagger.Connect(ctx, <options>)
if err != nil {
  panic(err)
}
defer c.Close()

// Do stuff with `c`
c.Container().From(...)
```

Since client options are no longer really needed (`LogOutput` is kinda deprecated by TUI, `Workdir` is made absolute by the new `Host` API), what if client creation was implicit?

```go
// Do stuff with `dagger` directly
dagger.Container().From(...)
```

In the entrypoint case, the call would be:

```go
func (e EngineTargets) Lint(ctx context.Context) (string, error) {
	return dagger.Container().
		From("golangci/golangci-lint:v1.51").
		WithMountedDirectory("/app", e.SrcDir).
		WithWorkdir("/app").
		WithExec([]string{"golangci-lint", "run", "-v", "--timeout", "5m"}).
		Stdout(ctx)
}
```

To make it work, client initialization can happen in two ways:

- **Lazy init**: There's a singleton client in the library that gets initialized when needed
- **Contextual**: A pre-init'ed client can be available in context, in which case the SDK blindly re-uses it. This is the path for TUI and extensions, which initialize an external client

/cc @helderco

Reply:
Yep this is what I commented on https://github.com/dagger/dagger/pull/5060#discussion_r1187705753 👍

Reply:
> Yep this is what I commented on [#5060 (comment)](https://github.com/dagger/dagger/pull/5060#discussion_r1187705753) 👍

I did see that comment, but it was still referencing `ctx.client` and `self.client`, not sure if I missed some part?

What I'm suggesting is we don't expose client. SDK can be used "directly" without going through a client (e.g. `dagger.Container()` instead of `client.Container()`), SDK initializes a client on-demand.

Somewhat similar to the `http` library in Go: you can create a `http.Client`, or you can use `http.Get` which uses the default client instance under the hood.

Reply:
Yeah in my recent pushes I removed `ConfigPath` and `NoExtensions` from the engine config (not needed anymore, all the project loading is in the API and done by `dagger do` using the Go SDK), so those are already gone.

I think this is a great idea, but it will be important to still leave explicit client creation as an option. Mostly just for our own internal needs where we need to, e.g., setup multiple engines and then connect to each of them simultaneously (happens in remotecache tests). That doesn't have to be anything nice-to-use or obvious, just available.

Reply:
> when would you close the connection?

We could add a `dagger.Close()`. With entrypoints it can be managed automatically.

Since we don't pass a context around in Python and Node.js, it would have to be the _lazy init_ option.